### DISCUSSION TITLE
What if you can't use block_until_ready?

### DISCUSSION THREAD

**Discussion by ntlm1686** [jax-ml/jax#10076] [Q&A] [answered]
I am trying to benchmark some code. However, it seems I can't use ```block_until_ready``` everywhere.

Is this a good way to benchmark this code?

```python
@jit
def polyval_gpu():
    return [jnp.polyval(r, 0.1) for r in SOME_DATA]

%timeit polyval_gpu()
```
OR
```python
@jit
def polyval_gpu():
    for r in SOME_DATA:
        ans = jnp.polyval(r, 0.1)
    return ans

%timeit polyval_gpu()
```

**Comment by jakevdp** [comment, marked as answer]
For non-array outputs, rather than calling the `array.block_until_ready()` method, you can call the `jax.block_until_ready()` function:
```python
%timeit jax.block_until_ready(polyval_gpu())
```

**Reply by ntlm1686** [reply]
Are these non-array outputs moved from GPU to CPU after calculated?
For example, in the code below
```python
@jit
def polymul_many_times_gpu():
    return [jnp.polymul(p, p) for p in P1]
```

**Reply by YouJiacheng** [reply]
> Are these non-array outputs moved from GPU to CPU after calculated? For example, in the code below
> 
> ```python
> @jit
> def polymul_many_times_gpu():
>     return [jnp.polymul(p, p) for p in P1]
> ```

No, unless you print them.

**Reply by jakevdp** [reply]
No, `block_until_ready()` does not materialize outputs to different devices, it simply blocks Python program execution until those arrays are finished computing.

**Reply by ntlm1686** [reply]
@jakevdp @YouJiacheng 
Thank you very much. I think my question has been solved.

**Comment by YouJiacheng** [comment]
`jax.block_until_ready` will `tree_map` `array.block_until_ready()` for you.
But you still cannot(and should not) use `block_until_ready` inside `jit`, you should use it on the return value of jitted function.
BTW, useless computations are likely(almost always) optimized by JIT.
```python
@jit
def polyval_gpu():
    for r in SOME_DATA:
        ans = jnp.polyval(r, 0.1)
    return ans

%timeit jax.block_until_ready(polyval_gpu())
```
`polyval_gpu` may only evaluate ` jnp.polyval(r, 0.1)` at the last data in SOME_DATA, thus the performance measured may be unreliable.

**Reply by ntlm1686** [reply]
Thank you very much.
